src/daemon/http/request.rs

```diff
@@ -0,0 +1,527 @@
+        }
+    }
+
+    /// Checks whether the request is a POST or returns an error response.
+    pub fn check_delete(&self) -> Result<(), HttpResponse> {
+        match *self.request.method() {
+            Method::DELETE => Ok(()),
+            _ => Err(HttpResponse::method_not_allowed()),
+        }
+    }
+
+    /*
+    /// Returns the full URI of the request.
+    pub fn uri(&self) -> &Uri {
+        self.request.uri()
+    }
+    */
+
+    /// Returns the current request path.
+    pub fn path(&self) -> Result<RequestPath, InvalidPath> {
+        RequestPath::from_request(self)
+    }
+
+    /// Returns a reference to the Hyper request.
+    pub fn hyper(&self) -> &HyperRequest {
+        &self.request
+    }
+
+    /*
+    /// Returns the headers of the request.
+    pub fn headers(&self) -> &HeaderMap {
+        self.request.headers()
+    }
+    */
+
+    /// Returns the user agent header if present.
+    pub fn user_agent(&self) -> Option<String> {
+        match self.request.headers().get(&USER_AGENT) {
+            None => None,
+            Some(value) => value.to_str().ok().map(|s| {
+                // Note: HeaderValue.to_str() only returns ok in case the
+                // value is plain       ascii so it's safe to
+                // treat bytes as characters here.
+                if s.len() > HTTP_USER_AGENT_TRUNCATE {
+                    s[..HTTP_USER_AGENT_TRUNCATE].to_string()
+                } else {
+                    s.to_string()
+                }
+            }),
+        }
+    }
+
+    /// Checks for permissions.
+    ///
+    /// Returns an appropriate error response if the permissions are not met.
+    pub fn check_permission(
+        &self, permission: Permission, resource: Option<&MyHandle>
+    ) -> Result<(), HttpResponse> {
+        self.auth.check_permission(permission, resource).map_err(|err| {
+            HttpResponse::response_from_error(Error::from(err))
+        })
+    }
+
+    /// Checks the permissions and progresses to the next processing stage.
+    ///
+    /// If the authentication information for the request has the given
+    /// permissions, returns an [`AuthedRequest`] and an [`Actor`] which
+    /// allow further processing.
+    ///
+    /// Otherwise returns an appropriate error response.
+    pub fn proceed_permitted(
+        self, 
+        permission: Permission,
+        resource: Option<&MyHandle>,
+    ) -> Result<(AuthedRequest<'a>, AuthInfo), HttpResponse> {
+        self.check_permission(permission, resource)?;
+        Ok((
+            AuthedRequest {
+                request: self.request,
+                server: self.server,
+                limits: self.limits,
+            },
+            self.auth
+        ))
+    }
+
+    /// Permits the request to the next processing stage.
+    ///
+    /// Returns [`AuthedRequest`] and [`Actor`] without requiring any
+    /// permissions whatsoever.
+    pub fn proceed_unchecked(
```

Comment:
When I read this at the call site I was confused, if it doesn't check anything why call it, what does it do? This is more like an `into_authenticated_reqyest()` fn. I can't think of a good name for it though. As it does the same as `proceed_permitted()` without the permission check, maybe an alternative could be to just call `check_permission()` from the call site followed by `into_authenticated_request()`.

Reply:
It can’t be called `into_authenticated_request` because that implies that authentication was checked. I want the three methods to have the same naming scheme to make it clear that you have to call one of the three to continue.